### ISSUE TITLE
2.0.5 - Lua\RegExp - SIGABRT (Thread 1 is about to kill the process)

### ISSUE THREAD

**Issue by hont512** [haproxy/haproxy#225]
I've described a bit in #211, looks like it is separate issue 

- Lua processing is too big to find the exact reason
- Lua settings in CFG are default
- It happens only in 2.0.2-2.0.5 (1.9.x is not affected)
Normally in 1.9.x I get PCRE MATCH_LIMIT exceptions sometimes. Maybe 2.0 has some changes in exceptions processing?

```
12:49:55 haproxy  Thread 1 is about to kill the process.
12:49:55 haproxy  *>Thread 1 : act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
12:49:55 haproxy               stuck=1 fdcache=0 prof=0 harmless=0 wantrdv=0
12:49:55 haproxy               cpu_ns: poll=18479063028 now=21064392998 diff=2585329970
12:49:55 haproxy               curr_task=0x55aac5d81db0 (task) calls=8 last=0
12:49:55 haproxy                 fct=0x55aac464baa0 (process_stream) ctx=0x55aaccf1c790
12:49:55 haproxy               strm=0x55aaccf1c790 src=1.2.3.4 fe=http-in be=http-in dst=unknown
12:49:55 haproxy               rqf=20808002 rqa=64030 rpf=80000000 rpa=0 sif=EST,200028 sib=INI,30
12:49:55 haproxy               af=(nil),0 csf=0x55aac56f8820,200
12:49:55 haproxy               ab=(nil),0 csb=(nil),0
12:49:55 haproxy               cof=0x55aac5c5bb20,80201306:PASS(0x55aace268c30)/RAW((nil))/tcpv4(529)
12:49:55 haproxy               cob=(nil),0:NONE((nil))/NONE((nil))/NONE(0)
```

```
gdb /usr/sbin/haproxy ./core.1000.haproxy.28478
GNU gdb (Debian 8.2.1-2) 8.2.1
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/sbin/haproxy...done.
[New LWP 28478]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -sf 27636'.
Program terminated with signal SIGABRT, Aborted.
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) thread apply all bt

Thread 1 (Thread 0x7effa826a140 (LWP 28478)):
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007effa841a535 in __GI_abort () at abort.c:79
#2  0x000055aac476190e in ha_panic () at src/debug.c:169
#3  0x000055aac4761c37 in wdt_handler (sig=<optimized out>, si=<optimized out>, arg=<optimized out>) at src/wdt.c:123
#4  <signal handler called>
#5  0x00007effa85d08fe in ?? () from /lib/x86_64-linux-gnu/libpcre.so.3
#6  0x00007effa85d5563 in ?? () from /lib/x86_64-linux-gnu/libpcre.so.3
#7  0x00007effa85d33bc in ?? () from /lib/x86_64-linux-gnu/libpcre.so.3
#8  0x00007effa85e2b59 in pcre_exec () from /lib/x86_64-linux-gnu/libpcre.so.3
#9  0x00007effa82401d9 in gmatch_exec () from /usr/local/lib/lua/5.3/rex_pcre.so
#10 0x00007effa824246d in generic_find_func () from /usr/local/lib/lua/5.3/rex_pcre.so
#11 0x00007effa8865585 in ?? () from /usr/lib/x86_64-linux-gnu/liblua5.3.so.0
#12 0x00007effa8872365 in ?? () from /usr/lib/x86_64-linux-gnu/liblua5.3.so.0
#13 0x00007effa8864d82 in ?? () from /usr/lib/x86_64-linux-gnu/liblua5.3.so.0
#14 0x00007effa8865a35 in lua_resume () from /usr/lib/x86_64-linux-gnu/liblua5.3.so.0
#15 0x000055aac46035ca in hlua_ctx_resume (yield_allowed=<optimized out>, lua=<optimized out>) at src/hlua.c:1107
#16 hlua_ctx_resume (lua=0x55aaccf3b5b0, yield_allowed=1) at src/hlua.c:1075
#17 0x000055aac460c80e in hlua_action (rule=0x55aac6bcf490, px=0x55aac6bbaeb0, sess=<optimized out>, s=0x55aaccf1c790, flags=2) at src/hlua.c:6859
#18 0x000055aac46199be in http_req_get_intercept_rule (px=px@entry=0x55aac6bbaeb0, rules=<optimized out>, s=s@entry=0x55aaccf1c790, deny_status=deny_status@entry=0x7ffc631a6e2c) at src/proto_http.c:1786
#19 0x000055aac461fb1f in http_process_req_common (s=s@entry=0x55aaccf1c790, req=req@entry=0x55aaccf1c7a0, an_bit=an_bit@entry=16, px=<optimized out>) at src/proto_http.c:2549
#20 0x000055aac464d293 in process_stream (t=t@entry=0x55aac5d81db0, context=0x55aaccf1c790, state=<optimized out>) at src/stream.c:2090
#21 0x000055aac47398dd in process_runnable_tasks () at src/task.c:412
#22 0x000055aac4692684 in run_poll_loop () at src/haproxy.c:2516
#23 run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:2637
#24 0x000055aac45e1647 in main (argc=<optimized out>, argv=<optimized out>) at src/haproxy.c:3314
```

**Comment by wtarreau** [comment]
Do you have an excessively big regex file ? Here what the trace says is that haproxy was completely frozen for 2.5 seconds doing exclusively this, which means dropping connections etc in parallel... Or maybe one of your regex is vulnerable and is being attacked. I have already seen some incorrectly written regex take multiple DAYS to evaluate. Usually such regex contain multiple groups that are ambiguous and which force the regex engine to evaluate many possibilities. For example:

  $ time echo | egrep '.+.{1,2500}'
  real    0m5.159s
  user    0m5.132s
  sys     0m0.024s

$ time echo | egrep '(.?){0,2000}'
  real    0m3.187s
  user    0m3.112s
  sys     0m0.072s

But I don't remember having seen this with PCRE. I don't think there's a way to find the original plain-text regex from the gmatch_exec() context, that would definitely help in such a situation.

**Comment by hont512** [comment]
I have a lot of extremely crazy regexps in 15mb Lua-file. It is very likely that the problem appears when some bad regexp meets bad data.  
But this ticket is created because there are no such problems in 1.9.x and it happens only in 2.0.

Thoughts
- The problem has different error-log in 1.9 (not likely)
- 1.9 wait more instead of killing all processes - I don't see it + real damage much less
- There are some other related differencies between 1.9 and 2.0

**Comment by wtarreau** [comment]
Oh the difference is trivial: you already had the problem in 1.9 except that 1.9 was not able to detect that it was frozen. 2.0 now has a watchdog to kill haproxy when it enters an endless loop (deadlock between threads, but somewhere, broken linked list etc). Spending one millisecond in a task is already a lot, spending 2 seconds is really not possible. It causes absurd bug reports such as flapping servers, dropped connections, 100% CPU etc which in fact are all byproducts of the process being 100% frozen for multiple seconds. There are places where in the time it takes you to evaluate one regex, people would have popped up a container, started haproxy, processed incoming requests, stopped it and stopped the container!

But even with a 15mB Lua file, Lua automatically yields to give the CPU back to haproxy periodically, so it's not related to your Lua file size at all. You really have something there (possibly a regex as in the examples above) which freezes haproxy for a very long time. I really don't know if there are ways to build Lua to support debuggers so that you can figure where it was in a core, that would be terribly helpful in such a case.

**Comment by hont512** [comment]
- Watchdog is the explanation, it is clear now. Thank you.

- For my case current watchdog acts like regression.
 
"2 seconds for some processing is a critical lua-side problem" - I get it of cause. 

Moreover, I have a big history of different problems with all that mess. That's why I use own "watchdog" that restarts Haproxy much faster than 2.5sec if it doesn't respond for new requests. 

As I understand, for "nbproc > 1" one stucked-for-few-seconds process is very big but not always-deathwish problem. So, in 1.9 my damage from these rare buggy delays is not too big. 
While 2.0 kills everything and makes it much more significant (by the way, reload operation for my config takes ~2 seconds, so this kill affects many current connections that would succeed other way + 2 seconds of new connections).

My config is buggy and now I'll start thinking how to check everything. But it is not something outstandingly bad. Approximately I have less than 30 bad requests from ~300 millions daily spreaded among a lot of servers and processes. Mistakes of such level might be not too rare among users. So, each such mistake would cost more real damage since 2.0. Of cause there should be some opposite cases where watchdog helps.

This ticket may be closed. But maybe some params to configure watchdog could save a lot.

**Comment by wtarreau** [comment]
Well, for now the watchdog has only helped, and actually way more than
initially expected, and fortunately apparently nobody else is facing
such extreme bad cases as yours. For now I really don't want to disable
the watchdog because I know how people think, they'll say "hmmm this
could crash my process, better disable it". Or at least I'd like to
find a very offensive option name such as "I-am-totally-stupid-and-I-definitely-want-to-run-into-serious-trouble-so-please-oh-please-disable-the-watchdog-I-promise-I-will-never-issue-a-bug-report"
to force people to think twice before doing it. And probably it would emit
warnings every time the watchdog would have fired, as a reminder that the
situation is not normal. So for the time being I prefer to continue without
such an option and just observe if that it helps some people fix a rare
bogus config or if we need to find more info to help (such as in your case
as you cannot yet figure the bogus regex, which is painful).

**Comment by wtarreau** [comment]
Please have a look there:

   https://zeux.io/2010/11/07/lua-callstack-with-c-debugger/

It indicates how to retrieve Lua line numbers from a core file. I think
I'll have to try to reproduce such a crash, verify that it works and
document it. There's also a gdb script here:

   https://stackoverflow.com/questions/8528503/how-can-i-get-the-lua-stack-trace-from-a-core-file-using-gdb

Please check if you can get something out of these using your core file.

**Comment by hont512** [comment]
- Watchdog helped even me by showing this ) 

It is almost ultimate good for single thread mode. But with multi-threading only smarter watchdog can act good way all the time (that is more difficult than ability to disable it). 

So, even when I'd solve these current problems, I still wouldn't be able to agree. 

Because forced kill-action is just about one simple thing: 
"Policy to kill every process when only one has problem _can_ increase damage"

For example, I develop DNS-backend with multiple worker-threads, sharing same socket. And policy to restart only one thread instead of full application in case of problems has saved me a lot of requests. Now I don't have problems there, but I know that if I make new rare bug in future or someone would find a way to attack and raise exceptions with buggy requests - there will be the lowest possible damage.

* By the way, if "I-am-totally-stupid-but-dont-kill-me-please" option is a support problem, but you'd decide someday to add it, it may be in build-config to be hidden from most users. 

- Thank you for propositions. I've tried GDB-script - it is not working "as is" and I'm not too familiar with GDB to figure it out at the moment. 

- "bt full" showed me even HTTP-headers from probably related request, hower I'm not sure if it has really buggy headers or it just looks this way in trace
```
#9  0x00007effa82401d9 in gmatch_exec () from /usr/local/lib/lua/5.3/rex_pcre.so
No symbol table info available.
#10 0x00007effa824246d in generic_find_func () from /usr/local/lib/lua/5.3/rex_pcre.so
No symbol table info available.
#11 0x00007effa8865585 in ?? () from /usr/lib/x86_64-linux-gnu/liblua5.3.so.0
No symbol table info available.
#12 0x00007effa8872365 in ?? () from /usr/lib/x86_64-linux-gnu/liblua5.3.so.0
No symbol table info available.
#13 0x00007effa8864d82 in ?? () from /usr/lib/x86_64-linux-gnu/liblua5.3.so.0
No symbol table info available.
#14 0x00007effa8865a35 in lua_resume () from /usr/lib/x86_64-linux-gnu/liblua5.3.so.0
No symbol table info available.
#15 0x000055aac46035ca in hlua_ctx_resume (yield_allowed=<optimized out>, lua=<optimized out>) at src/hlua.c:1107
        ret = <optimized out>
        msg = <optimized out>
        trace = <optimized out>
        ret = <optimized out>
        msg = <optimized out>
        trace = <optimized out>
        __pl_r = <optimized out>
--Type <RET> for more, q to quit, c to continue without paging--bt full c
        __pl_l = <optimized out>
        ret = <optimized out>
#16 hlua_ctx_resume (lua=0x55aaccf3b5b0, yield_allowed=1) at src/hlua.c:1075
        ret = <optimized out>
        msg = <optimized out>
        trace = <optimized out>
        __pl_r = <optimized out>
        __pl_l = <optimized out>
        ret = <optimized out>
#17 0x000055aac460c80e in hlua_action (rule=0x55aac6bcf490, px=0x55aac6bbaeb0, sess=<optimized out>, s=0x55aaccf1c790, flags=2) at src/hlua.c:6859
        arg = <optimized out>
        hflags = <optimized out>
        dir = 0
        error = <optimized out>
        msg = {size = 0, area = 0x0, data = 0, head = 0}
#18 0x000055aac46199be in http_req_get_intercept_rule (px=px@entry=0x55aac6bbaeb0, rules=<optimized out>, s=s@entry=0x55aaccf1c790, deny_status=deny_status@entry=0x7ffc631a6e2c)
    at src/proto_http.c:1786
        sess = 0x55aacef16690
        txn = <optimized out>
        rule = 0x55aac6bcf490
        ctx = {
          line = 0x55aad358fe8a "Accept-Encoding: gzip, deflate\r\nCache-Control: no-cache\r\nHost: xxxxxxxxx.xx\r\n\r\nhe-Control: no-cache\r\n\r\n\300z\316g\005\001]9\201m\203\244b\022P\232\317>e\226\301<\377", idx = 4, val = 30, vlen = 0, tws = 0, del = 30, prev = 4}
        auth_realm = <optimized out>
        early_hints = <optimized out>
        rule_ret = HTTP_RULE_RES_CONT
        act_flags = 2
        len = <optimized out>
#19 0x000055aac461fb1f in http_process_req_common (s=s@entry=0x55aaccf1c790, req=req@entry=0x55aaccf1c7a0, an_bit=an_bit@entry=16, px=<optimized out>) at src/proto_http.c:2549
        sess = <optimized out>
        txn = 0x55aac5a92d40
        msg = 0x55aac5a92da0
        rule = <optimized out>
        wl = <optimized out>
        verdict = <optimized out>
        deny_status = 2
        conn = 0x55aac5c5bb20
```

- The problem is not only about the case when watchdog triggers. 
I'm thinking to just replace regex-functions in LUA with custom wrappers that will log and measure performance.

- The biggest part of problem is that it is impossible to "check if the particular regex is good". 
Performance depends on processed string and anyway it is vulnerable. I will fix current regexps, but some day strange data would come and watchdog would start massacre again )

ModSecurity fights it with low PcreMatchLimit and PcreMatchLimitRecursion options that are like 1500 instead of 10kk defaults in pcre-lib. Some Lua-modules for web-servers have similar settings that affects all regexps in Lua-processing (https://github.com/openresty/lua-nginx-module#lua_regex_match_limit). 
It could be useful for haproxy too.

**Comment by wtarreau** [comment]
Note, I figured how to produce a Lua stack trace, it will be quite helpful for you as it will point to the exact call place. Now for example you'll get this (look at the end for the Lua functions and line):

*>Thread 1 : act=1 glob=0 wq=0 rq=0 tl=1 tlsz=0 rqsz=0
             stuck=1 fdcache=1 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=5213981 now=2011705496 diff=2006491515
             curr_task=0x21d12f0 (task) calls=1 last=0
               fct=0x482310 (process_stream) ctx=0x21dc5b0
             strm=0x21dc5b0 src=127.0.0.1 fe=http be=http dst=unknown
             rqf=908002 rqa=30 rpf=80000000 rpa=0 sif=EST,200028 sib=INI,30
             af=(nil),0 csf=0x21d5010,200
             ab=(nil),0 csb=(nil),0
             cof=0x21d1110,201300:PASS(0x21d12c0)/RAW((nil))/tcpv4(29)
             cob=(nil),0:NONE((nil))/NONE((nil))/NONE(0)
             Current executing Lua from a stream analyser -- stack traceback:
                lua-regex.lua:6: in function 'loop'
                lua-regex.lua:17: in function <lua-regex.lua:15>

Concerning the pcre-lib limits that's quite useful of course. However, unless I missed something, there's hardly anything that can be done about it in haproxy since the lua-pcre binding is totally external (the one I found was lrexlib and despite having successfully installed it, I still haven't figured how to use it as it complains that I cannot load dynamic libs).

Regarding the principle of killing all vs one, I disagree :-)  Here we're talking about threads. Every single time we've had to deal with a thread deadlock, all threads were deadlocked because sooner or later one thread enters a contention point with the looping thread. Also a sick thread shares the same memory space as others and even if you would kill an offending thread, it still leaves half-work started behind it. Your method is valable only for processes, provided they all serve the exact same service (which is not necessarily the case for us, nor is it for threads by the way since lots of people love to bind SSL to certain threads only for example).

I've merged this into master and backported it to 2.0, so if you pull latest 2.0 you'll have it already.

**Comment by hont512** [comment]
- Thank you. In future that LUA stack trace would save me a lot.

- I use lrexlib (+ native LUA patterns but very rare). 
```
luarocks install lrexlib-pcre
```
```
rex = require "rex_pcre"
local str = "ABCD"
if (rex.find(str,"B.D")) then
        core.Info("MATCH")
end
```

- I was going to re-compile pcre-lib already, but it is possible to set limits directly in the pattern like:
```
rex.find(str,"(*LIMIT_MATCH=10000)(*LIMIT_RECURSION=15000)B.D")
rex.find(str,"(*LIMIT_MATCH=10000)B.D")
rex.find(str,"(*LIMIT_RECURSION=15000)B.D")
```
Also, I've asked devs, looks like there will be better way to do it soon

- By the way, I didn't see such limits for native haproxy regexps. 
It takes just to add PCRE_EXTRA_MATCH_LIMIT and PCRE_EXTRA_MATCH_LIMIT_RECURSION in pcre_extra part of pcre_exec()

- I can't argue with you about haproxy internals =) 

However: 
1. The point is "it _can_ increase damage". It is confirmed. And this case won't be the one. 
Maybe somewhere I have huge 0.2-second regex and it is not fatal problem now. But it will become mass shooting when DDOS come and bring overload.

2. I thought we talk about a bunch of very independent processes. 

I don't know details about contention points for nbproc-mode. 
But before I have bugs when some processes became forever-stucked because of some socket-transfer problem during reload. 
I can't be sure, but statistics felt like I loose not more 10% requests if 1 of 10 processes is stucked (it was the case when watchdog could be really useful). 

If it is correct, in case of 10-sec stucked process I'll have 
\- Watchdog: 100% lost current requests (including not just 9 other good processes, but also some old processes that left from reloads) + 100% lost new requests during ~2 seconds restart (not sure, but I think OS would send back RST as for closed port until full haproxy initialization and visitors would get error in browser immediately)
5.5 second happiness left. If there will be no stuck again.

\- No watchdog: 
\-\- Extra delays: <10% current requests and probably 10% new requests
\-\- If there will be no TCP-activity at all, then very small amount of lost requests will start appear only after 3 seconds. 
Most of 10% new requests that should fail might switch process after TCP-retransmit. 
Many current requests that are already on buggy process won't notice even delay just because waiting for backend answer could be same or longer. 
Completely lost requests - few percent tops.

And I think it is the worst case while it is possible that stucked process just won't get new requests. 

3. That's why now I can use 2.0 only as very very expensive way to find most problematic cases

**Comment by hont512** [comment]
> lots of people love to bind SSL to certain threads only for example

And I think there are a lot of other cases when per-process\per-thread restarts are not possible or difficult to implement. 

It is just more reason why one-for-everyone solution may have problems. 
Haproxy is great and it used by huge amount very different people in very different conditions. 

Watchdog that can really suit 100% cases must be very smart - adjustable timeouts, optional per-process restarts, even advanced process checks to cover above SSL-case. 
It is not proposition. I'm strictly against about wasting your time in such way. While build-level kill\just-log switch also covers 100% cases )

**Comment by wtarreau** [comment]
On Wed, Aug 21, 2019 at 01:49:43PM -0700, hont512 wrote:
> - I use lrexlib (+ native LUA patterns but very rare). 

that's what I figured (though I tried lrexlib-pcre2 but that's the same).

> - I was going to re-compile pcre-lib already, but it is possible to set limits directly in the pattern like:
> ```
> rex.find(str,"(*LIMIT_MATCH=10000)(*LIMIT_RECURSION=15000)B.D")
> rex.find(str,"(*LIMIT_MATCH=10000)B.D")
> rex.find(str,"(*LIMIT_RECURSION=15000)B.D")
> ```
> Also, I've asked devs, looks like there will be better way to do it soon

In case it would support environment variables, you can set them in your
global section, so it could be useful to check.

> - By the way, I didn't see such limits for native haproxy regexps. 
> It takes just to add PCRE_EXTRA_MATCH_LIMIT and PCRE_EXTRA_MATCH_LIMIT_RECURSION in pcre_extra part of pcre_exec()

OK, we can have a look. However people generally don't abuse regex in
haproxy since they often just need to match prefixes or suffixes, which
are available without regex.

> - I can't argue with you about haproxy internals =) 
> 
> However: 
> 1. The point is "it _can_ increase damage". It is confirmed. And this case won't be the one. 
> Maybe somewhere I have huge 0.2-second regex and it is not fatal problem now.
> But it will become mass shooting when DDOS come and bring overload.

Note that if you have an operation that takes as long as 200ms and it's
accessible outside, your LB is deemed to be attacked, and even a simple
browser or curl can render it totally unusable. I mean, with only 5
requests per second, it will not respond to any traffic at all anymore.

> 2. I thought we talk about a bunch of very independent processes. 
> 
> I don't know details about contention points for nbproc-mode. 
> But before I have bugs when some processes became forever-stucked because of
> some socket-transfer problem during reload. 

No they are not "stuck", the just remain alive until the connection ends
or the "hard-stop-after" timeout strikes. But these processes have no
other impact on the service.

> I can't be sure, but statistics felt like I loose not more 10% requests if 1
> of 10 processes is stucked (it was the case when watchdog could be really
> useful). 

+ potentially manual intervention to stop the problem. And if your processes
don't all take the same traffic, it can be 10% of the services which are
totally inaccessible until you act.

> If it is correct, in case of 10-sec stucked process I'll have 
> \- Watchdog: 100% lost current requests (including not just 9 other good
> processes, but also some old processes that left from reloads)

No, it will not touch the previous processes if they're working fine. It
will only kill the process which is frozen.

> + 100% lost
> new requests during ~2 seconds restart (not sure, but I think OS would send
> back RST as for closed port until full haproxy initialization and visitors
> would get error in browser immediately)

And how do you compare this to 100% frozen for 10 seconds, one minute, or
even one hour, until someone decides that the service still seems to accept
connections (just accepted at the system level), but never processes them ?

> 5.5 second happiness left. If there will be no stuck again.
> 
> \- No watchdog: 
> \-\- Extra delays: <10% current requests and probably 10% new requests

It's 10% only if you're dealing with processes (in which case the watchdog
would have only killed the faulty process), but with threads, it becomes
100% as soon as anything requires synchronization (server state update,
access to the CLI, etc). Also the long pause will have for effect to
trigger server state changes due to apparent timeouts, which can cause
a domino effect forcing the 10% to quickly become 100%.

> \-\- If there will be no TCP-activity at all, then very small amount of lost
> requests will start appear only after 3 seconds. 

And all transfers in progress that time-out on the browser or on the server.

> Most of 10% new requests that should fail might switch process after
> TCP-retransmit. 

No, they are in the TCP accept queue and are delivered to a given listener
queue assigned to a process. Thus they are not redispatched.

> Many current requests that are already on buggy process won't notice even
> delay just because waiting for backend answer could be same or longer. 

It just depends on the delay and how often it repeats. Your servers might
very well cut transfers after 10s of lack of progress. Browsers will abort
as well after some time. If you have a few such requests in parallel, the
delays add up and no progress happens.

> Completely lost requests - few percent tops.

I think you did not yet realize that we're publicly discussing the fact
that your site is affected by a major DoS vulnerability by which a single
user can block 100% of the traffic by sending one request every 3 seconds.
If anyone figures the site, finds the request, and figures the regex, it
only requires one request every 3s to make your site 100% unavailable
forever. So you're arguing about percents of availability for a situation
which should not even happen in the first place and that you were not even
aware of before you had the chance to be notified about.

At least with the latest patches you'll get the information about where
the problem happens in the Lua code. Maybe it's always the same place,
and you'll fix it quickly. Maybe you'll figure a class of bad regex that
need to be addressed. And maybe you'll figure the bad pattern and will
be able to block it with a few ACLs before calling Lua. This could be
the simplest solution in the short term.

**Comment by hont512** [comment]
> And how do you compare this to 100% frozen for 10 seconds, one minute, or
even one hour, until someone decides that the service still seems to accept
connections (just accepted at the system level), but never processes them ?

I've never argued that it is helpful in many cases and even described the one I have ) 
It is just about "probability of case when it won't be more than 10 second, than damage is bigger". In most other cases it is less. 

> No, it will not touch the previous processes if they're working fine. It
will only kill the process which is frozen.

In this way I wouldn't cry so much for lost requests= ) And there wouldn't be such noticeable hole in my network-load graph

It was opposite in my case. 

This is my last crash. That time it was "nbproc 6" and you can see amount of terminated processes. A bit bigger than one.

```
12:49:55 haproxy  Thread 1 is about to kill the process.
12:49:55 haproxy  *>Thread 1 : act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
12:49:55 haproxy               stuck=1 fdcache=0 prof=0 harmless=0 wantrdv=0
12:49:55 haproxy               cpu_ns: poll=18479063028 now=21064392998 diff=2585329970
12:49:55 haproxy               curr_task=0x55aac5d81db0 (task) calls=8 last=0
12:49:55 haproxy                 fct=0x55aac464baa0 (process_stream) ctx=0x55aaccf1c790
12:49:55 haproxy               strm=0x55aaccf1c790 src=1.2.3.4 fe=http-in be=http-in dst=unknown
12:49:55 haproxy               rqf=20808002 rqa=64030 rpf=80000000 rpa=0 sif=EST,200028 sib=INI,30
12:49:55 haproxy               af=(nil),0 csf=0x55aac56f8820,200
12:49:55 haproxy               ab=(nil),0 csb=(nil),0
12:49:55 haproxy               cof=0x55aac5c5bb20,80201306:PASS(0x55aace268c30)/RAW((nil))/tcpv4(529)
12:49:55 haproxy               cob=(nil),0:NONE((nil))/NONE((nil))/NONE(0)
12:49:58 haproxy  [ALERT] 230/124958 (5357) : Current worker #6 (28478) exited with code 134 (Aborted)
12:49:58 haproxy  [ALERT] 230/124958 (5357) : exit-on-failure: killing every processes with SIGTERM
12:49:58 haproxy  [ALERT] 230/124958 (5357) : Current worker #1 (28471) exited with code 143 (Terminated)
12:49:58 haproxy  [ALERT] 230/124958 (5357) : Current worker #2 (28472) exited with code 143 (Terminated)
12:49:58 haproxy  [ALERT] 230/124958 (5357) : Current worker #3 (28474) exited with code 143 (Terminated)
12:49:58 haproxy  [WARNING] 230/124958 (5357) : Former worker #1 (27636) exited with code 143 (Terminated)
12:49:58 haproxy  [WARNING] 230/124958 (5357) : Former worker #2 (26729) exited with code 143 (Terminated)
12:49:58 haproxy  [WARNING] 230/124958 (5357) : Former worker #1 (26728) exited with code 143 (Terminated)
12:49:58 haproxy  [WARNING] 230/124958 (5357) : Former worker #2 (27637) exited with code 143 (Terminated)
12:49:58 haproxy  [ALERT] 230/124958 (5357) : Current worker #4 (28475) exited with code 143 (Terminated)
12:49:58 haproxy  [WARNING] 230/124958 (5357) : Former worker #3 (26731) exited with code 143 (Terminated)
12:49:58 haproxy  [WARNING] 230/124958 (5357) : Former worker #3 (27638) exited with code 143 (Terminated)
12:49:58 haproxy  [WARNING] 230/124958 (5357) : Former worker #4 (26732) exited with code 143 (Terminated)
12:49:58 haproxy  [WARNING] 230/124958 (5357) : Former worker #4 (27639) exited with code 143 (Terminated)
12:49:58 haproxy  [ALERT] 230/124958 (5357) : Current worker #5 (28476) exited with code 143 (Terminated)
12:49:58 haproxy  [WARNING] 230/124958 (5357) : Former worker #6 (26734) exited with code 143 (Terminated)
12:49:58 haproxy  [WARNING] 230/124958 (5357) : Former worker #5 (27640) exited with code 143 (Terminated)
12:49:58 haproxy  [WARNING] 230/124958 (5357) : Former worker #6 (27641) exited with code 143 (Terminated)
12:49:58 haproxy  [WARNING] 230/124958 (5357) : Former worker #5 (26733) exited with code 143 (Terminated)
12:49:58 haproxy  [WARNING] 230/124958 (5357) : All workers exited. Exiting... (134)
12:49:58 systemd  haproxy.service: Main process exited, code=exited, status=134/n/a
12:49:58 systemd  haproxy.service: Failed with result 'exit-code'.
12:49:59 systemd  haproxy.service: Service RestartSec=100ms expired, scheduling restart.
12:49:59 systemd  haproxy.service: Scheduled restart job, restart counter is at 1.
12:49:59 systemd  Stopped HAProxy Load Balancer.
12:49:59 systemd  Starting HAProxy Load Balancer...
12:50:00 mywatchdog  HAPROXY 1st problem | ERR Failed to connect to 127.0.0.1 port 80: Connection refused | CONNECT_TIME 0 | TIME 0.101
12:50:00 mywatchdog  HAPROXY 2nd problem - forcing RESTART | ERR Failed to connect to 127.0.0.1 port 80: Connection refused | CONNECT_TIME 0 | TIME 0.046
```

Now I see that it's also not compatible with my watchdog which make 2nd restart because of errors during haproxy-initiated 1st restart )

**Comment by hont512** [comment]
Oooups... Looks like I should set "master-worker no-exit-on-failure". And then damage will be like 20 times less, that is much more acceptable.

**Comment by wtarreau** [comment]
Ah if you're on multi-process via the mworker I understand how you lost more than one process. With this said, really, please update to latest 2.0-git so that you at least know what's happening in your Lua code. It's really dangerous to stay exposed to the net with such a major vulnerability in your code that you have not even precisely located yet.

**Comment by hont512** [comment]
I've tried with "no-exit-on-failure". 
\- The problem was in a part of regexp with one mistake ([^z]*\\.x). As expected it is exactly the case "will be free in 10 seconds or so" ) For most data performance is not fatal, only rare requests raise execution time to seconds. 
\- Two processes were killed in few seconds (client retried request after first kill). Now damage difference is not so big, but still killing is more expensive when this DoS vulnerability is used randomly. In case if it will be used intentionally - anyway zero chances to survive. 

I have big amount of another places with same regex and many similar problems. Just forcing match-limits is necessary and acceptable but not too good solution itself. In addition to it I hope to find some analyzer-api with abilities like https://regex101.com to check steps-count for different data and find at least some obviously bad cases in advance. 

Thank you extremely very super much for this discussion and watchdog =)  I've expected regex-problems for long time but since I get some MATCH_LIMIT errors, I thought that such significant delays are impossible. And I've spent phenomenal amount of time fighting with some problems that were related with it. Now it was huge lifesaver for me.

**Comment by wtarreau** [comment]
You're welcome. Any viable long-term solution to address match limits in regex would be nice, especially if a generic solution can be enforced. I'm pretty sure you're now having a big incentive in digging in this area, please do not hesitate to share your findings. May I close the issue now ?

**Comment by hont512** [comment]
Lrexlib is going to be updated. When it happens, I'll write together with other findings. 
I think there won't be universal way and code for each regexp should be changed, but maybe at least it will be a useful note in management-doc. 

Issue may be closed

**Comment by wtarreau** [comment]
Yep! Thank you, now closing.
